src/compiler/checker.ts

```diff
@@ -10403,6 +10433,7 @@ namespace ts {
                 case SyntaxKind.TemplateExpression:
                     return checkTemplateExpression(<TemplateExpression>node);
                 case SyntaxKind.StringLiteral:
+                    return checkStringLiteralExpression(<LiteralExpression>node);
                 case SyntaxKind.NoSubstitutionTemplateLiteral:

```

Comment:
I don't think the original issue was so terrible. You only ran into this case when you were contextually typed by a type with a string literal constituent to begin with. In practice, this is not frequently encountered.

However, @jbondc, I think you'll like the the current implementation a lot better (and for various reasons). We now only create a string literal type if the constituent types have a string literal type, but not just if their content is equal.

Furthermore, the types are cached in a map of strings to string literal types. Whenever testing the assignability of two string literal types, reference equality kicks in (which is fast).

Additionally, error messages are slightly better because you have a specific literal type to report (i.e. `Type '"Foo"' is not assignable to '"Bar"'.` instead of `Type 'string' is not assignable to '"Bar"'.`).